src/puppetlabs/trapperkeeper/bootstrap.clj

```diff
@@ -132,26 +136,102 @@
+  [protocol-id :- schema/Keyword
+   duplicate-services :- [(schema/protocol services/ServiceDefinition)]
+   service->entry-map :- {(schema/protocol services/ServiceDefinition) AnnotatedBootstrapEntry}]
+  (let [make-error-message (fn [service]
+                             (let [entry (get service->entry-map service)]
+                               (format "%s:%s\n%s"
+                                       (:bootstrap-file entry)
+                                       (:line-number entry)
+                                       (:entry entry))))]
+    (IllegalArgumentException.
+      (format (str "Duplicate implementations found for service protocol '%s':\n%s")
+              protocol-id
+              (string/join "\n" (map make-error-message duplicate-services))))))
+
+(schema/defn check-duplicate-service-implementations
+  "Throws an exception if two services implement the same service protocol"
+  [services :- [(schema/protocol services/ServiceDefinition)]
+   bootstrap-entries :- [AnnotatedBootstrapEntry]]
+
+  ; Zip up the services and bootstrap entries and construct a map out of them
+  ; to use as a lookup table below
+  (let [service->entry-map (zipmap services bootstrap-entries)]
+    ; Find duplicates base on the service id returned by calling service-def-id
+    ; on each service
+    (if-let [duplicate (first (find-duplicates services services/service-def-id))]
+      (throw (duplicate-protocol-error (key duplicate) (val duplicate) service->entry-map)))))
+
+(schema/defn bootstrap-error :- IllegalArgumentException
+  [entry :- schema/Str
+   bootstrap-file :- schema/Str
+   line-number :- schema/Int
+   original-message :- schema/Str]
+  (IllegalArgumentException.
+    (format (str "Problem loading service '%s' on line '%s' in bootstrap "
+                 "configuration file '%s':\n%s")
+            entry line-number bootstrap-file original-message)))
+
+(schema/defn resolve-services!
+  [bootstrap-entries]
+  (for [{:keys [bootstrap-file line-number entry]} bootstrap-entries]
+    (try+
+      (let [{:keys [namespace service-name]} (parse-bootstrap-line! entry)]
+        (resolve-service! namespace service-name ))
+      ; Catch and re-throw as java exception
+      (catch [:type :illegal-argument] {:keys [message]}
+        (throw (bootstrap-error entry bootstrap-file line-number message))))))
+
+(schema/defn remove-duplicate-entries :- [AnnotatedBootstrapEntry]
+  "Removes duplicate entries by only looking at the :entry key in each entry map.
+   This way, duplicate entries in different files and on different lines are removed"
+  [entries :- [AnnotatedBootstrapEntry]]
+  ; Construct a [k v] pair where the key is the text of the bootstrap entry, and the
+  ; value is the original AnnotatedBootstrapEntry map. By shoving them into a map,
+  ; duplicate keys are ignored. We pick out the values with vals, and get back our
+  ; original list minus the duplicates.
+  (vals (into {} (for [entry entries] [(:entry entry) entry]))))
```

Comment:
I don't like magical one liners, but this was the easiest to read method I could come up with. Happy to take suggestions.